memory/cache_capacity_based_memory_allocator.h

```diff
@@ -0,0 +1,71 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#pragma once
+
+#include <mutex>
+
+#include "cache/cache_reservation_manager.h"
+#include "utilities/memory_allocators.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+// CacheCapacityBasedMemoryAllocator is for allocating memory based on available
+// cache space by charging the memory to the cache.
+// For now, it does not support the actual memory allocation and only the
+// memory charging.
```

Comment:
[Open for discussion]
For BlockBasedTableReader, memory usages are allocated somewhere outside of this Allocator and tracking-only is sufficient. Therefore I didn't impl the conventional Allocate/Deallocate function. But for future use-case of `CacheCapacityBasedMemoryAllocator`, we might (?) need actual allocation.

Reply:
OK, so it turns out we are reserving based on `BlockBasedTable::ApproximateMemoryUsage()`, not making (de)allocations go through the `Allocator` interface. Do you think we should just use the CacheReservationManager interface directly (with appropriate locking)?

Reply:
> not making (de)allocations go through the Allocator interface

Reply:
> Do you think we should just use the CacheReservationManager interface directly (with appropriate locking)?

Yeah .. I've been thinking about this too.  This task deepened my understanding toward Allocator approach - the difficulties of it lies in going through explicit (de)allocation through our `Allocator` interface for objects whose de(allocation) are handles automatically (like [new_table_properties->user_collected_properties.insert({key, raw_val.ToString()});](https://github.com/facebook/rocksdb/blob/7.0.fb/table/meta_blocks.cc#L336-L337)).
If we were not able to achieve that, then "Allocator" is essentially a tracker.


> CacheReservationManager interface directly (with appropriate locking)?

I am thinking of making `CacheReservationManager` optionally internal thread-safe (`new CacheReservationManager(cache,  true \*enable_internal_locking*\)`). "Internal thread-safe" because it's easier to do so than passing the lock pointer around if we were to do so for this PR without Allocator layer. "Optionally" is because the `WriteBufferManager` maintains a atomic-relaxed `memory_used_` that serves both as a stat for query without locking (trade accuracy for speed) and a referenced data point for charging and releasing cache entry (requires accuracy) and I have no plan to deal with the refactoring issue there (double locking and deadlock) for now.

Reply:
> It seems achievable but could take a long time. For example we could make STL use our allocator like [7b1890f0b3e096a43006e4163b98f6316ac1e91b](https://github.com/ClickHouse/ClickHouse/blob/5e641e2dace33180ec34eb913deaaebd2ad01c18/src/Common/AllocatorWithMemoryTracking.h), 

It's true - I looked into that route too but I agreed with what you said about "take a long time" but worth exploring when we have time and are bored :).


> Can you make a thread-safe wrapper or something?
Aha - wrapper also works (just haven't thought of that)!

Reply:
Done